actors/market/src/state.rs

```diff
@@ -567,6 +568,95 @@ impl State {
         Ok(rval_pending_deal)
     }
 
+    /// Delete proposal and state simultaneously.
+    pub fn remove_completed_deal<BS>(
+        &mut self,
+        store: &BS,
+        deal_id: DealID,
+    ) -> Result<(), ActorError>
+    where
+        BS: Blockstore,
+    {
+        let deleted = self.remove_deal_state(store, deal_id)?;
+        if deleted.is_none() {
+            return Err(actor_error!(illegal_state, "failed to delete deal state: does not exist"));
+        }
+
+        let deleted = self.remove_proposal(store, deal_id)?;
+        if deleted.is_none() {
+            return Err(actor_error!(
+                illegal_state,
+                "failed to delete deal proposal: does not exist"
+            ));
+        }
+
+        Ok(())
+    }
+
+    /// Given a DealProposal, checks that the corresponding deal has activated
+    /// If not, checks that the deal is past its activation epoch and performs cleanup
+    pub fn get_active_deal_or_process_timeout<BS>(
+        &mut self,
+        store: &BS,
+        curr_epoch: ChainEpoch,
+        deal_id: DealID,
+        deal_proposal: &DealProposal,
+        dcid: &Cid,
+    ) -> Result<(Option<DealState>, TokenAmount), ActorError>
+    where
+        BS: Blockstore,
+    {
+        let deal_state = self.find_deal_state(store, deal_id)?;
+
+        match deal_state {
+            Some(deal_state) => Ok((Some(deal_state), TokenAmount::zero())),
+            None => {
+                // deal was not activated in time
+
+                // deal_id called too early
+                if curr_epoch < deal_proposal.start_epoch {
+                    return Err(actor_error!(
```

Comment:
This causes an abort if someone tries to settle a deal that was published, but hasn't been activated yet. It would be nicer if this performed a no-op settlement instead. We can return `None` rather than `Err` here, but must also return a flag to distinguish this from the expired case, so the caller doesn't report the deal as expired.